r/adbcdrivermanager/src/async.cc

```diff
@@ -0,0 +1,262 @@
+    UNPROTECT(2);
+
+    // Set up the call to run the callback
+    SEXP callback_sym = PROTECT(Rf_install("callback"));
+    SEXP return_code_sexp = PROTECT(Rf_ScalarInteger(callback.return_code));
+    SEXP callback_call =
+        PROTECT(Rf_lang4(callback_sym, return_code_sexp, error_xptr, return_value_xptr));
+
+    // Run the callback
+    Rf_eval(callback_call, env_sexp);
+
+    UNPROTECT(6);
+  }
+
+  static void FinalizeXptr(SEXP xptr) {
+    CallbackQueue* queue = reinterpret_cast<CallbackQueue*>(R_ExternalPtrAddr(xptr));
+    delete queue;
+  }
+
+  static SEXP MakeXptr(SEXP shelter = R_NilValue) {
+    SEXP xptr = PROTECT(R_MakeExternalPtr(new CallbackQueue(), R_NilValue, shelter));
+    R_RegisterCFinalizer(xptr, &FinalizeXptr);
+    UNPROTECT(1);
+    return xptr;
+  }
+
+ private:
+  std::deque<RCallback> pending_callbacks_;
+  std::mutex callbacks_lock_;
+};
+
+// TODO: unlike other parts of this package, these functions make use of
+// C++ standard library functions that might throw exceptions in functions
+// that are directly called from R. These will segfault if an exception is
+// thrown. cpp11 handles this using BEGIN_CPP11 an END_CPP11...we would need
+// similar for safety here.
+
+extern "C" SEXP RAdbcNewCallbackQueue(void) { return CallbackQueue::MakeXptr(); }
+
+extern "C" SEXP RAdbcCallbackQueueRunPending(SEXP callback_queue_xptr) {
+  // TODO: Check callback_queue_xptr class
+  auto queue = reinterpret_cast<CallbackQueue*>(R_ExternalPtrAddr(callback_queue_xptr));
+  return Rf_ScalarReal(queue->RunPending());
+}
+
+extern "C" SEXP RAdbcStatementExecuteQueryAsync(SEXP callback_queue_xptr,
+                                                SEXP statement_xptr, SEXP out_stream_xptr,
+                                                SEXP error_xptr, SEXP callback_env) {
+  // TODO: check array_stream/array/callback queue classes using utils in radbc.h
+  auto queue = reinterpret_cast<CallbackQueue*>(R_ExternalPtrAddr(callback_queue_xptr));
+  auto statement = adbc_from_xptr<AdbcStatement>(statement_xptr);
+  auto out_stream = adbc_from_xptr<ArrowArrayStream>(out_stream_xptr);
+  auto error = adbc_from_xptr<AdbcError>(error_xptr);
+
+  // Task handle to ensure the thread pointer is cleaned up
+  SEXP task_xptr = PROTECT(Task::MakeXptr(callback_queue_xptr));
+  SEXP task_symbol = PROTECT(Rf_install("task"));
+  Rf_setVar(task_symbol, task_xptr, callback_env);
+  UNPROTECT(1);
+
+  auto task = reinterpret_cast<Task*>(R_ExternalPtrAddr(task_xptr));
+  CallbackQueue::RCallback callback =
+      queue->InitCallback(callback_env, out_stream_xptr, error_xptr);
+  task->worker = new std::thread([statement, out_stream, error, callback, queue] {
+    CallbackQueue::RCallback callback_out = callback;
+    int64_t rows_affected = -1;
+    callback_out.return_code =
+        AdbcStatementExecuteQuery(statement, out_stream, &rows_affected, error);
+    queue->AddCallback(callback_out);
+  });
+
+  UNPROTECT(1);
+  return R_NilValue;
```

Comment:
Can we return `task_xptr` here, and in the R layer create a promise bound to this specific task?